### ISSUE TITLE
Upgrading Apigee module past v24 introduces breaking changes for no reason

### ISSUE THREAD

**Issue by cygnus8595** [GoogleCloudPlatform/cloud-foundation-fabric#1643]
[This commit](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/commit/a42463522169e70c293fab76f76b89104319a128)  introduced a few breaking changes that would cause arbitrary downtime for reasons that seem to be just for preference and not for any security reasons.

First change was in the instance naming. In [this line](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/commit/a42463522169e70c293fab76f76b89104319a128#diff-721af65e06c447c8a3d2ecd8afda4d11f88f8fb4069582c294363c5215222740L91), the name of the instance was changed to include an `instance-` prefix that forces re-creation of the instance and causes downtime for any production organizations that were created with a prior version of the module (in my case, version 20).

The second change was in the instance attachments. In [this block](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/commit/a42463522169e70c293fab76f76b89104319a128#diff-721af65e06c447c8a3d2ecd8afda4d11f88f8fb4069582c294363c5215222740L101), the `for_each` loop was changed from looping through the environments listed in the `instances` variable to looping through the regions in the `environments` variable. This introduces ambiguity in the module usage, as it was changed from a required variable to an optional variable. If there are no regions specified in the `environments` block, the instance attachments are destroyed and not re-created if the code isn't changed.

Looking at the commits, I don't see a reason why this was done except for personal preference, and I think it made the modules less usable by anyone not using the blueprints. 

I suggest that these changes should be reverted back to how they were in version 23 and older so that they don't create downtime or ambiguity.

Example of instance re-creation due to new name:
```
-/+ resource "google_apigee_instance" "instances" {
      ~ host                 = "10.19.204.2" -> (known after apply)
      ~ id                   = "organizations/tcna-nonprod-apigee/instances/us-east1" -> (known after apply)
      ~ name                 = "us-east1" -> "instance-us-east1" # forces replacement
      ~ peering_cidr_range   = "SLASH_22" -> (known after apply)
      ~ port                 = "443" -> (known after apply)
      ~ service_attachment   = "projects/jc4438b4093aea498p-tp/regions/us-east1/serviceAttachments/apigee-us-east1-gff4" -> (known after apply)
        # (5 unchanged attributes hidden)
    }
```
Example of instance attachment being destroyed and not re-created due to the change in the `for_each` block:
```
# module.apigee_platform.module.apigee.google_apigee_instance_attachment.instance_attachments["us-east1-ce-dev"] will be destroyed
  # (because key ["us-east1-ce-dev"] is not in for_each map)
  - resource "google_apigee_instance_attachment" "instance_attachments" {
      - environment = "ce-dev" -> null
      - id          = "organizations/tcna-nonprod-apigee/instances/us-east1/attachments/6a421321-3242-4b10-ab54-edb87622f0fc" -> null
      - instance_id = "organizations/tcna-nonprod-apigee/instances/us-east1" -> null
      - name        = "6a421321-3242-4b10-ab54-edb87622f0fc" -> null
    }
```

**Comment by juliocc** [comment]
I'm not the author of the PR but my understanding is that the refactor was done because you can have at most one instance per region. Given that restriction, I think the PR is fine and  not motivated by personal preference of the author.

That being said, we can probably work around the two issues you point out. The first one can be easily fixed by letting the user override the instance name. The second issue can be addressed using moved blocks to tell terraform the names of the new resources.

We can guide you through the process of creating a PR if you're up to it.

**Comment by cygnus8595** [comment]
Thanks @juliocc , I'd be interested in getting a PR created. Let me know how to proceed.

**Comment by juliocc** [comment]
@cygnus8595 start by reading the [quickstart](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#quick-developer-workflow) in our contributing guide, which will explain how to 
- Fork this repository
- Make your changes in your forked repository
- [Open a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request?tool=webui)

You want to make two changes:

**1. Allow instances to have user defined name:** 
- modify the `instances` variable and a new optional `name` field of type string
- Use this field (if non-null) for the instance name in the resource, which will end up looking something like `name = coalesce(each.value.name, "instance-${each.key}")`

**2. Include `moved` blocks**
I'm not sure exactly how to do it. I suggest you check the `moved` blocks [documentation](https://developer.hashicorp.com/terraform/language/modules/develop/refactoring) and do a few tests. Do let me know if you find any issues

**Comment by juliocc** [comment]
@cygnus8595 I'm closing this. The first issue was addressed by #1653 and the other one is clearly marked in the release notes as an incompatible change. 

Feel free to open a PR to add the moved blocks as discussed.
